src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+    //
+    cmpOp1->gtRequestSetFlags();
+#endif
+
+    if (!optReturnBlock)
+    {
+        // Update edges if b1: BBJ_COND and b2: BBJ_COND
+
+        flowList* edge1 = comp->fgGetPredForBlock(b1->bbJumpDest, b1);
+        flowList* edge2;
+
+        if (sameTarget)
+        {
+            edge2 = comp->fgGetPredForBlock(b2->bbJumpDest, b2);
+        }
+        else
+        {
+            edge2 = comp->fgGetPredForBlock(b2->bbNext, b2);
+
+            comp->fgRemoveRefPred(b1->bbJumpDest, b1);
+
+            b1->bbJumpDest = b2->bbJumpDest;
+
+            comp->fgAddRefPred(b2->bbJumpDest, b1);
+        }
+
+        noway_assert(edge1 != nullptr);
+        noway_assert(edge2 != nullptr);
+
+        BasicBlock::weight_t edgeSumMin = edge1->edgeWeightMin() + edge2->edgeWeightMin();
+        BasicBlock::weight_t edgeSumMax = edge1->edgeWeightMax() + edge2->edgeWeightMax();
+        if ((edgeSumMax >= edge1->edgeWeightMax()) && (edgeSumMax >= edge2->edgeWeightMax()))
+        {
+            edge1->setEdgeWeights(edgeSumMin, edgeSumMax, b1->bbJumpDest);
+        }
+        else
+        {
+            edge1->setEdgeWeights(BB_ZERO_WEIGHT, BB_MAX_WEIGHT, b1->bbJumpDest);
+        }
+    }
+
+    /* Modify the target of the conditional jump and update bbRefs and bbPreds */
+
+    if (optReturnBlock)
+    {
+        comp->fgRemoveRefPred(b1->bbJumpDest, b1);
+
+        noway_assert(!b2->bbJumpDest);
+        b1->bbJumpDest = b2->bbJumpDest;
+        b1->bbJumpKind = b2->bbJumpKind;
+        b1->bbJumpSwt  = b2->bbJumpSwt;
+
+        noway_assert(b1->bbJumpKind == BBJ_RETURN);
+        noway_assert(b2->bbJumpKind == BBJ_RETURN);
+        noway_assert(b1->bbNext == b2);
+        noway_assert(b2->bbNext == b3);
+        noway_assert(b3);
+    }
+    else
+    {
+        noway_assert(b1->bbJumpKind == BBJ_COND);
+        noway_assert(b2->bbJumpKind == BBJ_COND);
+        noway_assert(b1->bbJumpDest == b2->bbJumpDest);
+        noway_assert(b1->bbNext == b2);
+        noway_assert(b2->bbNext);
+    }
+
+    // Get rid of the second block
+
+    comp->fgUnlinkBlock(b2);
+    b2->bbFlags |= BBF_REMOVED;
+    // If b2 was the last block of a try or handler, update the EH table.
+    comp->ehUpdateForDeletedBlock(b2);
+
+    if (optReturnBlock)
+    {
+        // Get rid of the third block
+        comp->fgUnlinkBlock(b3);
+        b3->bbFlags |= BBF_REMOVED;
+        // If b3 was the last block of a try or handler, update the EH table.
+        comp->ehUpdateForDeletedBlock(b3);
+    }
+
+    if (!optReturnBlock)
+    {
+        // Update bbRefs and bbPreds
+        //
+        // Replace pred 'b2' for 'b2->bbNext' with 'b1'
+        // Remove  pred 'b2' for 'b2->bbJumpDest'
+        comp->fgReplacePred(b2->bbNext, b2, b1);
+        comp->fgRemoveRefPred(b2->bbJumpDest, b2);
+    }
+
+    // Update loop table
+    comp->fgUpdateLoopsAfterCompacting(b1, b2);
+    if (optReturnBlock)
+    {
+        comp->fgUpdateLoopsAfterCompacting(b1, b3);
```

Comment:
The comment for `fgUpdateLoopsAfterCompacting` says that (for these arguments), `b1->bbNext == b3`. That's not true here. But maybe the comments for that function should be updated? It might not matter (and this call might not do anything) because I don't think a BBJ_RETURN `b3` block would ever be a distinguished block in a loop.

Reply:
TODO: 
`comp->fgUpdateLoopsAfterCompacting(b1, b2);` was there for original code.
To check if `comp->fgUpdateLoopsAfterCompacting(b1, b3);` updates the loop table.

Reply:
Created https://github.com/dotnet/runtime/issues/55605 and added this item there.